R/getInsertSizeDistFromBam.R

```diff
@@ -0,0 +1,108 @@
+#' @title Tabulate insert sizes from paired-end alignments in bam files.
+#'
+#' @description Read and tabulate the insert sizes from paired-end alignments
+#'   contained in one or several bam files. By default, all properly aligned
+#'   read pairs are included. Optionally, alignments can be restricted to
+#'   those in a specific genomic region region (\code{regions} argument) or
+#'   the number of alignments read can be limited (\code{nmax} argument).
+#'
+#' @author Michael Stadler
+#'
+#' @param fname \code{character} vector with one or several bam files. If
+#'   multiple files are given, insert sizes from all will be pooled and
+#'   tabulated together.
+#' @param regions \code{GRanges} object. Only alignments falling into these
+#'   regions will be used. If \code{NULL} (the default), all alignments are
+#'   used.
+#' @param nmax \code{numeric(1)} specifying the maximal number of alignments to
+#'   read. If \code{NA} (the default), the alignments in \code{regions} (if
+#'   \code{regions} are not \code{NULL}) or in the bam file will be used.
+#' @param isizemax \code{numeric(1)} specifying the maximal insert size to
+#'   report. Larger insert sizes will be set to \code{isizemax} with a warning.
+#'
+#' @return \code{integer} vector with the number of insert sizes. The element at
+#'   position \code{i} gives the observed number of alignment pairs with an
+#'   insert size of \code{i}. The number of insert sizes greater than
+#'   \code{isizemax} that were set to \code{isizemax} are reported in the
+#'   attribute \code{"ncapped"}.
+#'
+#' @seealso \code{\link[Rsamtools]{scanBam}} used to read alignments.
+#'
+#' @examples
+#' if (requireNamespace("Rsamtools", quietly = TRUE)) {
+#'     bamf <- system.file("extdata", "getInsertSizeDistFromBam", "atac_mm10.bam",
+#'                         package = "swissknife")
+#'     isize <- getInsertSizeDistFromBam(bamf)
+#'     attr(isize, "ncapped")
+#'     plot(isize, type = "l",
+#'          xlab = "Insert size (bp)", ylab = "Number of fragments")
+#' }
+#'
+#' @importFrom IRanges IRangesList
+#' 
+#' @export
+getInsertSizeDistFromBam <- function(fname,
+                                     regions = NULL,
+                                     nmax = NA_integer_,
+                                     isizemax = 800) {
+    .assertPackagesAvailable(c("Rsamtools"))
+    if (!is.null(regions) && !inherits(regions, "GRanges")) {
+        stop("'regions' must be either NULL or a GRanges object.")
+    }
+    if (!is.na(nmax) && (!is.numeric(nmax) || length(nmax) != 1L || nmax <= 0)) {
+        stop("'nmax' must be either NA or a numeric scalar greater than zero.")
+    }
+    if (!is.numeric(isizemax) || length(isizemax) != 1L || isizemax <= 0) {
+        stop("'isizemax' must be a numeric scalar greater than zero.")
+    }
```

Comment:
would it be worth to have the option that the max can be the actual maximum fragment size in a specific bam?

Reply:
I think this would be currently hard to implement (I am initialising a vector with the counts, and to know its length I would therefore have to traverse the bam file first, and then again to get the counts. Also, there are almost always a couple of (very likely artefactual) alignments with huge insert sizes (>100kB), which would not make much sense to include. I would prefer to leave it as it is. You can always set `isizemax` to a large value if you need that.

Reply:
I agree, that makes sense.